libs/ml/examples/distributed_learning/main.cpp

```diff
@@ -0,0 +1,420 @@
+      Test(loss);
+
+      // Save loss variation data
+      // Upload to https://plot.ly/create/#/ for visualisation
+      if (lossfile)
+      {
+        lossfile << GetTimeStamp() << ", " << loss << "\n";
+      }
+    }
+
+    lossfile << GetTimeStamp() << ", "
+             << "STOPPED"
+             << "\n";
+    lossfile.close();
+  }
+
+  /**
+   * Train one batch
+   * @return training batch loss
+   */
+  DataType Train()
+  {
+    dataloader_.SetMode(fetch::ml::dataloaders::DataLoaderMode::TRAIN);
+    dataloader_.SetRandomMode(true);
+
+    DataType loss        = 0;
+    bool     is_done_set = false;
+
+    std::pair<TensorType, std::vector<TensorType>> input;
+    input = dataloader_.PrepareBatch(batch_size_, is_done_set);
+
+    {
+      std::lock_guard<std::mutex> l(model_mutex_);
+
+      g_.SetInput("Input", input.second.at(0));
+      g_.SetInput("Label", input.first);
+
+      TensorType loss_tensor = g_.ForwardPropagate("Error");
+      loss                   = *(loss_tensor.begin());
+      g_.BackPropagateError("Error");
+    }
+
+    return loss;
+  }
+
+  /**
+   * Run model on test set to get test loss
+   * @param test_loss
+   */
+  void Test(DataType &test_loss)
+  {
+    dataloader_.SetMode(fetch::ml::dataloaders::DataLoaderMode::TEST);
+
+    // Disable random to run model on whole test set
+    dataloader_.SetRandomMode(false);
+
+    SizeType test_set_size = dataloader_.Size();
+
+    dataloader_.Reset();
+    bool is_done_set;
+    auto test_pair = dataloader_.PrepareBatch(test_set_size, is_done_set);
+    {
+      std::lock_guard<std::mutex> l(model_mutex_);
+
+      g_.SetInput("Input", test_pair.second.at(0));
+      g_.SetInput("Label", test_pair.first);
+
+      test_loss = *(g_.Evaluate("Error").begin());
+    }
+  }
+
+  /**
+   * @return vector of gradient update values
+   */
+  TensorVectorType GetGradients() const
+  {
+    std::lock_guard<std::mutex> l(const_cast<std::mutex &>(model_mutex_));
+    return g_.GetGradients();
+  }
+
+  /**
+   * @return vector of weights that represents the model
+   */
+  TensorVectorType GetWeights() const
+  {
+    std::lock_guard<std::mutex> l(const_cast<std::mutex &>(model_mutex_));
+    return g_.get_weights();
+  }
+
+  /**
+   * Add pointers to other clients
+   * @param clients
+   */
+  void AddPeers(std::vector<std::shared_ptr<TrainingClient>> const &clients)
+  {
+    for (auto const &p : clients)
+    {
+      if (p.get() != this)
+      {
+        peers_.push_back(p);
+      }
+    }
+  }
+
+  /**
+   * Adds own gradient to peers queues
+   */
+  void BroadcastGradients()
+  {
+    // Load own gradient
+    TensorVectorType current_gradient = g_.GetGradients();
+
+    // Give gradients to peers
+    for (unsigned int i(0); i < number_of_peers_; ++i)
+    {
+      peers_[i]->AddGradient(current_gradient);
+    }
+  }
+
+  /**
+   * Adds gradient to own gradient queue
+   * @param gradient
+   */
+  void AddGradient(TensorVectorType gradient)
+  {
+    {
```

Comment:
is this pair of braces necessary? presumably the mutex goes out of scope immediately after the function call anyway?